docs/src/tutorials/Mixed-integer linear programs/facility_location.jl

```diff
@@ -0,0 +1,264 @@
+# Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors    #src
+# This Source Code Form is subject to the terms of the Mozilla Public License   #src
+# v.2.0. If a copy of the MPL was not distributed with this file, You can       #src
+# obtain one at https://mozilla.org/MPL/2.0/.                                   #src
+
+# # Facility Location
+
+# It was originally contributed by Mathieu Tanneau (@mtanneau) and
+# Alexis Montoison (@amontoison).
+
+using Random
+using LinearAlgebra
+
+using JuMP
+import GLPK
+using Plots
+
+# ## Uncapacitated facility location
+
+# ### Problem description
+# 
+# We are given
+# * A set $M=\{1, \dots, m\}$ of clients
+# * A set $N=\{ 1, \dots, n\}$ of sites where a facility can be built
+# 
+# **Decision variables**
+# Decision variables are split into two categories:
+# * Binary variable $y_{j}$ indicates whether facility $j$ is built or not
+# * Binary variable $x_{i, j}$ indicates whether client $i$ is assigned to facility $j$
+# 
+# **Objective**
+# The objective is to minimize the total cost of serving all clients.
+# This costs breaks down into two components:
+# * Fixed cost of building a facility.
+# In this example, this cost is $f_{j} = 1, \ \forall j$.
+# * Cost of serving clients from the assigned facility.
+# In this example, the cost $c_{i, j}$
+# of serving client $i$ from facility $j$
+# is the Euclidean distance between the two.
+# 
+# **Constraints**
+# * Each customer must be served by exactly one facility
+# * A facility cannot serve any client unless it is open
+
+# ### MILP formulation
+# 
+# The problem can be formulated as the following MILP:
+# 
+# ```math
+# \begin{aligned}
+# \min_{x, y} \ \ \ &
+# \sum_{i, j} c_{i, j} x_{i, j} + 
+# \sum_{j} f_{j} y_{j} \\
+# s.t. &
+# \sum_{j} x_{i, j} = 1, && \forall i \in M \\
+# & x_{i, j} \leq y_{j}, && \forall i \in M, j \in N \\
+# & x_{i, j}, y_{j} \in \{0, 1\}, && \forall i \in M, j \in N
+# \end{aligned}
+# ```
+# 
+# where the first set of constraints ensures
+# that each client is served exactly once,
+# and the second set of constraints ensures
+# that no client is served from an unopened facility.
+
+# ### Problem data
+
+
+Random.seed!(314)
+
+## number of clients
+m = 12
+## number of facility locations
+n = 5
+
+## Clients' locations
+Xc = rand(m)
+Yc = rand(m)
+
+## Facilities' potential locations
+Xf = rand(n)
+Yf = rand(n)
+
+## Fixed costs
+f = ones(n);
+
+## Distance
+c = zeros(m, n)
+for i in 1:m
+    for j in 1:n
+        c[i, j] = norm([Xc[i] - Xf[j], Yc[i] - Yf[j]], 2)
+    end
+end
+
+# Display the data
+scatter(Xc, Yc, label = "Clients", markershape=:circle, markercolor=:blue)
+scatter!(Xf, Yf, label="Facility", 
+    markershape=:square, markercolor=:white, markersize=6,
+    markerstrokecolor=:red, markerstrokewidth=2
+)
```

Comment:
I should run the formatter over these tutorials...

Reply:
yes! https://github.com/jump-dev/JuMP.jl/issues/2659